### PR TITLE
[FLOC-3691] Add an invariant for `providedBy`.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-3691

### CODE REVIEW

File: flocker/common/_interface.py [hunk 1]
```diff
@@ -54,3 +54,27 @@ def class_decorator(cls):
...
+
+    :param InterfaceClass interface: The interface to check for.
+
+    :return: A function that takes an object and returns a tuple of `bool` and
+        a error message.
+    """
+    interface_name = interface.__name__
+
+    def invariant(value):
+        if interface.providedBy(value):
```
Comment (sarum90) [inline flocker/common/_interface.py]: :grey_question: Is there a reason not to just do:

```
return (interface.providedBy(value), "{value!r} doesn't provide {interface}".format(
            value=value, interface=interface_name,
        ))
```

The API seems to suggest that's the more standard way to do it:
http://pyrsistent.readthedocs.org/en/latest/intro.html#invariants

File: flocker/common/test/test_interface.py [hunk 2]
```diff
@@ -132,3 +133,54 @@ def test_raise(self, logger):
...
+        """
+        self.assertRaises(
+            InvariantException,
+            InterfaceHolder, value=object(),
+        )
+
+    def test_invariant_message(self):
+        """
+        When an object not providing the given interface is provided,
+        an ``InvariantException`` is raised.
```
Comment (sarum90) [inline flocker/common/test/test_interface.py]: :grey_question: docstring doesn't seem to be updated for this test method.

File: flocker/common/test/test_interface.py [hunk 3]
```diff
@@ -132,3 +133,54 @@ def test_raise(self, logger):
...
+        exception = self.assertRaises(
+            InvariantException,
+            InterfaceHolder, value=True,
+        )
+        self.assertEqual(
+            exception.invariant_errors,
+            ("True doesn't provide IDummy",),
+        )
+
+    def test_invariant_name(self):
```
Comment (sarum90) [inline flocker/common/test/test_interface.py]: :grey_question: Do we require a docstring on all test methods? it looks like we do, but I'm not sure what it adds in this case. Name is rather descriptive, test is simple... Fix if you desire to and just missed this one, I'm indifferent.

**Comment by sarum90** [comment]
That's a nice little utility :smile: 

I've got a couple optional docs-related queries for you to glance over. Address them or not at your discretion. Otherwise, merge when you believe the tests to be green-enough.

File: flocker/common/_interface.py (hunk 1 above)
Comment (tomprince) [inline flocker/common/_interface.py]: I don't want to pay the cost of generating the representation if it isn't needed.

Comment (tomprince) [inline flocker/common/_interface.py]: https://github.com/tobgu/pyrsistent/issues/75

File: flocker/common/test/test_interface.py (hunk 3 above)
Comment (tomprince) [inline flocker/common/test/test_interface.py]: We do, I just missed adding it. Thanks for pointing it out. :+1:

**Pull request merged**